db: TestIteratorErrors failed [iterator is valid with non-nil Error]

https://github.com/cockroachdb/pebble/actions/runs/7568342774/job/20609357007
```
=== RUN   TestIteratorErrors
    external_test.go:31: Using seed 1705572529975016370
```

```
--- FAIL: TestIteratorErrors (1.79s)
panic: pebble: iterator is valid with non-nil Error [recovered]
	panic: pebble: iterator is valid with non-nil Error

goroutine 121985 [running]:
testing.tRunner.func1.2({0xd72be0, 0xc00136b940})
	/opt/hostedtoolcache/go/1.21.5/x64/src/testing/testing.go:1545 +0x238
testing.tRunner.func1()
	/opt/hostedtoolcache/go/1.21.5/x64/src/testing/testing.go:1548 +0x397
panic({0xd72be0?, 0xc00136b940?})
	/opt/hostedtoolcache/go/1.21.5/x64/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/pebble.(*Iterator).Valid(0xc001e22800?)
	/home/runner/work/pebble/pebble/iterator.go:2241 +0x87
github.com/cockroachdb/pebble/metamorphic.(*retryableIter).withRetry(0xc0034426c0, 0xc000ca1960)
	/home/runner/work/pebble/pebble/metamorphic/retryable.go:57 +0x66
github.com/cockroachdb/pebble/metamorphic.(*retryableIter).SeekGE(0xc003529820?, {0xc002071e37?, 0xc002720090?, 0xe09e58?})
	/home/runner/work/pebble/pebble/metamorphic/retryable.go:147 +0x5e
github.com/cockroachdb/pebble/metamorphic.(*iterSeekGEOp).run(0xc004390c40, 0x4131c5?, {0xc003529820?, 0xd3f980?, 0xc002720100?})
	/home/runner/work/pebble/pebble/metamorphic/ops.go:1356 +0xd1
github.com/cockroachdb/pebble/metamorphic.(*Test).step(0xc000d76960, 0xc00022b340?, 0x2c?)
	/home/runner/work/pebble/pebble/metamorphic/test.go:352 +0x48
github.com/cockroachdb/pebble/metamorphic.(*Test).Step(0xc000d76960)
	/home/runner/work/pebble/pebble/metamorphic/test.go:341 +0x75
github.com/cockroachdb/pebble_test.TestIteratorErrors(0xc00057b380)
	/home/runner/work/pebble/pebble/external_test.go:86 +0x898
testing.tRunner(0xc00057b380, 0xe5c3d8)
	/opt/hostedtoolcache/go/1.21.5/x64/src/testing/testing.go:1595 +0xff
created by testing.(*T).Run in goroutine 1
	/opt/hostedtoolcache/go/1.21.5/x64/src/testing/testing.go:1648 +0x3ad
```

Ran it with a change to print the verbose error on failure (to capture the stack of when the error was injected):

```
panic: pebble: backing file 003036 error: injected error
(1) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/pebble.(*tableCacheValue).load
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:1131
  | github.com/cockroachdb/pebble.(*tableCacheShard).findNodeInternal.func2
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:883
  | runtime/pprof.Do
  | 	/usr/local/go/src/runtime/pprof/runtime.go:51
  | github.com/cockroachdb/pebble.(*tableCacheShard).findNodeInternal
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:882
  | github.com/cockroachdb/pebble.(*tableCacheShard).findNode
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:778
  | github.com/cockroachdb/pebble.(*tableCacheShard).newRangeKeyIter
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:596
  | github.com/cockroachdb/pebble.(*tableCacheContainer).newRangeKeyIter
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:150
  | github.com/cockroachdb/pebble/internal/keyspan.(*LevelIter).loadFile
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/level_iter.go:183
  | github.com/cockroachdb/pebble/internal/keyspan.(*LevelIter).skipEmptyFileForward
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/level_iter.go:415
  | github.com/cockroachdb/pebble/internal/keyspan.(*LevelIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/level_iter.go:345
  | github.com/cockroachdb/pebble/internal/keyspan.(*mergingIterLevel).next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:251
  | github.com/cockroachdb/pebble/internal/keyspan.(*MergingIter).nextEntry
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:1041
  | github.com/cockroachdb/pebble/internal/keyspan.(*MergingIter).findNextFragmentSet
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:865
  | github.com/cockroachdb/pebble/internal/keyspan.(*MergingIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:640
  | github.com/cockroachdb/pebble/internal/keyspan.(*BoundedIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/bounded.go:159
  | github.com/cockroachdb/pebble/internal/keyspan.(*DefragmentingIter).defragmentForward
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/defragment.go:441
  | github.com/cockroachdb/pebble/internal/keyspan.(*DefragmentingIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/defragment.go:357
  | github.com/cockroachdb/pebble/internal/keyspan.(*assertIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/assert_iter.go:147
  | github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).nextPos
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/interleaving_iter.go:681
  | github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).yieldPosition
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/interleaving_iter.go:807
  | github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/interleaving_iter.go:478
  | github.com/cockroachdb/pebble.(*Iterator).nextPointCurrentUserKey
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/iterator.go:654
  | github.com/cockroachdb/pebble.(*Iterator).findNextEntry
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/iterator.go:587
  | github.com/cockroachdb/pebble.(*Iterator).First
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/iterator.go:1623
  | github.com/cockroachdb/pebble/metamorphic.(*retryableIter).First.func1
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/metamorphic/retryable.go:73
  | github.com/cockroachdb/pebble/metamorphic.(*retryableIter).withRetry
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/metamorphic/retryable.go:47
  | github.com/cockroachdb/pebble/metamorphic.(*retryableIter).First
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/metamorphic/retryable.go:72
  | github.com/cockroachdb/pebble/metamorphic.(*iterFirstOp).run
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/metamorphic/ops.go:1474
  | github.com/cockroachdb/pebble/metamorphic.(*Test).step
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/metamorphic/test.go:352
  | github.com/cockroachdb/pebble/metamorphic.(*Test).Step
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/metamorphic/test.go:341
  | github.com/cockroachdb/pebble_test.TestIteratorErrors
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/external_test.go:85
Wraps: (2) pebble: backing file 003036 error
Wraps: (3) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/pebble/vfs/errorfs.LabelledError.MaybeError
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/vfs/errorfs/dsl.go:259
  | github.com/cockroachdb/pebble/vfs/errorfs.(*Counter).MaybeError
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/vfs/errorfs/errorfs.go:213
  | github.com/cockroachdb/pebble/vfs/errorfs.(*Toggle).MaybeError
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/vfs/errorfs/errorfs.go:257
  | github.com/cockroachdb/pebble/vfs/errorfs.(*errorFile).Stat
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/vfs/errorfs/errorfs.go:505
  | github.com/cockroachdb/pebble/objstorage/objstorageprovider.newFileReadable
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/objstorage/objstorageprovider/vfs_readable.go:37
  | github.com/cockroachdb/pebble/objstorage/objstorageprovider.(*provider).vfsOpenForReading
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/objstorage/objstorageprovider/vfs.go:34
  | github.com/cockroachdb/pebble/objstorage/objstorageprovider.(*provider).OpenForReading
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/objstorage/objstorageprovider/provider.go:226
  | github.com/cockroachdb/pebble.(*tableCacheValue).load
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:1123
  | github.com/cockroachdb/pebble.(*tableCacheShard).findNodeInternal.func2
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:883
  | runtime/pprof.Do
  | 	/usr/local/go/src/runtime/pprof/runtime.go:51
  | github.com/cockroachdb/pebble.(*tableCacheShard).findNodeInternal
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:882
  | github.com/cockroachdb/pebble.(*tableCacheShard).findNode
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:778
  | github.com/cockroachdb/pebble.(*tableCacheShard).newRangeKeyIter
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:596
  | github.com/cockroachdb/pebble.(*tableCacheContainer).newRangeKeyIter
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:150
  | github.com/cockroachdb/pebble/internal/keyspan.(*LevelIter).loadFile
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/level_iter.go:183
  | github.com/cockroachdb/pebble/internal/keyspan.(*LevelIter).skipEmptyFileForward
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/level_iter.go:415
  | github.com/cockroachdb/pebble/internal/keyspan.(*LevelIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/level_iter.go:345
  | github.com/cockroachdb/pebble/internal/keyspan.(*mergingIterLevel).next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:251
  | github.com/cockroachdb/pebble/internal/keyspan.(*MergingIter).nextEntry
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:1041
  | github.com/cockroachdb/pebble/internal/keyspan.(*MergingIter).findNextFragmentSet
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:865
  | github.com/cockroachdb/pebble/internal/keyspan.(*MergingIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:640
  | github.com/cockroachdb/pebble/internal/keyspan.(*BoundedIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/bounded.go:159
  | github.com/cockroachdb/pebble/internal/keyspan.(*DefragmentingIter).defragmentForward
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/defragment.go:441
  | github.com/cockroachdb/pebble/internal/keyspan.(*DefragmentingIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/defragment.go:357
  | github.com/cockroachdb/pebble/internal/keyspan.(*assertIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/assert_iter.go:147
  | github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).nextPos
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/interleaving_iter.go:681
  | github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).yieldPosition
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/interleaving_iter.go:807
  | github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/interleaving_iter.go:478
  | github.com/cockroachdb/pebble.(*Iterator).nextPointCurrentUserKey
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/iterator.go:654
  | github.com/cockroachdb/pebble.(*Iterator).findNextEntry
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/iterator.go:587
  | github.com/cockroachdb/pebble.(*Iterator).First
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/iterator.go:1623
  | github.com/cockroachdb/pebble/metamorphic.(*retryableIter).First.func1
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/metamorphic/retryable.go:73
Wraps: (4) injected error
Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) errorfs.LabelledError [recovered]
	panic: pebble: backing file 003036 error: injected error
(1) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/pebble.(*tableCacheValue).load
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:1131
  | github.com/cockroachdb/pebble.(*tableCacheShard).findNodeInternal.func2
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:883
  | runtime/pprof.Do
  | 	/usr/local/go/src/runtime/pprof/runtime.go:51
  | github.com/cockroachdb/pebble.(*tableCacheShard).findNodeInternal
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:882
  | github.com/cockroachdb/pebble.(*tableCacheShard).findNode
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:778
  | github.com/cockroachdb/pebble.(*tableCacheShard).newRangeKeyIter
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:596
  | github.com/cockroachdb/pebble.(*tableCacheContainer).newRangeKeyIter
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:150
  | github.com/cockroachdb/pebble/internal/keyspan.(*LevelIter).loadFile
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/level_iter.go:183
  | github.com/cockroachdb/pebble/internal/keyspan.(*LevelIter).skipEmptyFileForward
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/level_iter.go:415
  | github.com/cockroachdb/pebble/internal/keyspan.(*LevelIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/level_iter.go:345
  | github.com/cockroachdb/pebble/internal/keyspan.(*mergingIterLevel).next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:251
  | github.com/cockroachdb/pebble/internal/keyspan.(*MergingIter).nextEntry
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:1041
  | github.com/cockroachdb/pebble/internal/keyspan.(*MergingIter).findNextFragmentSet
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:865
  | github.com/cockroachdb/pebble/internal/keyspan.(*MergingIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:640
  | github.com/cockroachdb/pebble/internal/keyspan.(*BoundedIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/bounded.go:159
  | github.com/cockroachdb/pebble/internal/keyspan.(*DefragmentingIter).defragmentForward
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/defragment.go:441
  | github.com/cockroachdb/pebble/internal/keyspan.(*DefragmentingIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/defragment.go:357
  | github.com/cockroachdb/pebble/internal/keyspan.(*assertIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/assert_iter.go:147
  | github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).nextPos
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/interleaving_iter.go:681
  | github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).yieldPosition
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/interleaving_iter.go:807
  | github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/interleaving_iter.go:478
  | github.com/cockroachdb/pebble.(*Iterator).nextPointCurrentUserKey
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/iterator.go:654
  | github.com/cockroachdb/pebble.(*Iterator).findNextEntry
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/iterator.go:587
  | github.com/cockroachdb/pebble.(*Iterator).First
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/iterator.go:1623
  | github.com/cockroachdb/pebble/metamorphic.(*retryableIter).First.func1
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/metamorphic/retryable.go:73
  | github.com/cockroachdb/pebble/metamorphic.(*retryableIter).withRetry
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/metamorphic/retryable.go:47
  | github.com/cockroachdb/pebble/metamorphic.(*retryableIter).First
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/metamorphic/retryable.go:72
  | github.com/cockroachdb/pebble/metamorphic.(*iterFirstOp).run
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/metamorphic/ops.go:1474
  | github.com/cockroachdb/pebble/metamorphic.(*Test).step
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/metamorphic/test.go:352
  | github.com/cockroachdb/pebble/metamorphic.(*Test).Step
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/metamorphic/test.go:341
  | github.com/cockroachdb/pebble_test.TestIteratorErrors
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/external_test.go:85
Wraps: (2) pebble: backing file 003036 error
Wraps: (3) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/pebble/vfs/errorfs.LabelledError.MaybeError
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/vfs/errorfs/dsl.go:259
  | github.com/cockroachdb/pebble/vfs/errorfs.(*Counter).MaybeError
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/vfs/errorfs/errorfs.go:213
  | github.com/cockroachdb/pebble/vfs/errorfs.(*Toggle).MaybeError
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/vfs/errorfs/errorfs.go:257
  | github.com/cockroachdb/pebble/vfs/errorfs.(*errorFile).Stat
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/vfs/errorfs/errorfs.go:505
  | github.com/cockroachdb/pebble/objstorage/objstorageprovider.newFileReadable
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/objstorage/objstorageprovider/vfs_readable.go:37
  | github.com/cockroachdb/pebble/objstorage/objstorageprovider.(*provider).vfsOpenForReading
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/objstorage/objstorageprovider/vfs.go:34
  | github.com/cockroachdb/pebble/objstorage/objstorageprovider.(*provider).OpenForReading
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/objstorage/objstorageprovider/provider.go:226
  | github.com/cockroachdb/pebble.(*tableCacheValue).load
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:1123
  | github.com/cockroachdb/pebble.(*tableCacheShard).findNodeInternal.func2
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:883
  | runtime/pprof.Do
  | 	/usr/local/go/src/runtime/pprof/runtime.go:51
  | github.com/cockroachdb/pebble.(*tableCacheShard).findNodeInternal
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:882
  | github.com/cockroachdb/pebble.(*tableCacheShard).findNode
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:778
  | github.com/cockroachdb/pebble.(*tableCacheShard).newRangeKeyIter
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:596
  | github.com/cockroachdb/pebble.(*tableCacheContainer).newRangeKeyIter
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:150
  | github.com/cockroachdb/pebble/internal/keyspan.(*LevelIter).loadFile
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/level_iter.go:183
  | github.com/cockroachdb/pebble/internal/keyspan.(*LevelIter).skipEmptyFileForward
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/level_iter.go:415
  | github.com/cockroachdb/pebble/internal/keyspan.(*LevelIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/level_iter.go:345
  | github.com/cockroachdb/pebble/internal/keyspan.(*mergingIterLevel).next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:251
  | github.com/cockroachdb/pebble/internal/keyspan.(*MergingIter).nextEntry
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:1041
  | github.com/cockroachdb/pebble/internal/keyspan.(*MergingIter).findNextFragmentSet
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:865
  | github.com/cockroachdb/pebble/internal/keyspan.(*MergingIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:640
  | github.com/cockroachdb/pebble/internal/keyspan.(*BoundedIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/bounded.go:159
  | github.com/cockroachdb/pebble/internal/keyspan.(*DefragmentingIter).defragmentForward
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/defragment.go:441
  | github.com/cockroachdb/pebble/internal/keyspan.(*DefragmentingIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/defragment.go:357
  | github.com/cockroachdb/pebble/internal/keyspan.(*assertIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/assert_iter.go:147
  | github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).nextPos
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/interleaving_iter.go:681
  | github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).yieldPosition
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/interleaving_iter.go:807
  | github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).Next
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/internal/keyspan/interleaving_iter.go:478
  | github.com/cockroachdb/pebble.(*Iterator).nextPointCurrentUserKey
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/iterator.go:654
  | github.com/cockroachdb/pebble.(*Iterator).findNextEntry
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/iterator.go:587
  | github.com/cockroachdb/pebble.(*Iterator).First
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/iterator.go:1623
  | github.com/cockroachdb/pebble/metamorphic.(*retryableIter).First.func1
  | 	/Users/jackson/go/src/github.com/cockroachdb/pebble/metamorphic/retryable.go:73
Wraps: (4) injected error
Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) errorfs.LabelledError
```

--- a/cleaner.go
+++ b/cleaner.go
@@ -12,7 +12,6 @@ import (
 
 	"github.com/cockroachdb/errors/oserror"
 	"github.com/cockroachdb/pebble/internal/base"
-	"github.com/cockroachdb/pebble/internal/invariants"
 	"github.com/cockroachdb/pebble/objstorage"
 	"github.com/cockroachdb/tokenbucket"
 )
@@ -122,10 +121,6 @@ func (cm *cleanupManager) EnqueueJob(jobID int, obsoleteFiles []obsoleteFile) {
 	cm.mu.totalJobs++
 	cm.maybeLogLocked()
 	cm.mu.Unlock()
-
-	if invariants.Enabled && len(cm.jobsCh) >= cap(cm.jobsCh)-2 {
-		panic("cleanup jobs queue full")
-	}
 
 	cm.jobsCh <- job
 }

--- a/external_test.go
+++ b/external_test.go
@@ -62,7 +62,7 @@ func TestIteratorErrors(t *testing.T) {
 		testOpts.Opts.DisableAutomaticCompactions = true
 
 		// Create an errorfs injector that injects ErrInjected on 5% of reads.
-		// Wrap it in both a counter and a toggle so that we a) know wether an
+		// Wrap it in both a counter and a toggle so that we a) know whether an
 		// error was injected over the course of an operation, and b) so that we
 		// can disable error injection during Open.
 		predicate := errorfs.And(errorfs.Reads, errorfs.Randomly(0.50, seed))
@@ -72,9 +72,16 @@ func TestIteratorErrors(t *testing.T) {
 		testOpts.Opts.ReadOnly = true
 
 		test, err := metamorphic.New(
-			metamorphic.GenerateOps(rng, 500, metamorphic.ReadOpConfig()),
+			metamorphic.GenerateOps(rng, 5000, metamorphic.ReadOpConfig()),
 			testOpts, "" /* dir */, &testWriter{t: t})
 		require.NoError(t, err)
+
+		defer func() {
+			if r := recover(); r != nil {
+				t.Errorf("last injected error: %+v", counter.LastError())
+				panic(r)
+			}
+		}()
 
 		// Begin injecting errors.
 		toggle.On()
@@ -92,7 +99,7 @@ func TestIteratorErrors(t *testing.T) {
 			newCount := counter.Load()
 			if diff := newCount - prevCount; diff > 0 {
 				if !strings.Contains(operationOutput, errorfs.ErrInjected.Error()) {
-					t.Errorf("Injected %d errors in op %d but the operation output %q does not contain the injected error: %+v",
+					t.Fatalf("Injected %d errors in op %d but the operation output %q does not contain the injected error: %+v",
 						diff, i, operationOutput, counter.LastError())
 				}
 			}

--- a/get_iter.go
+++ b/get_iter.go
@@ -73,6 +73,10 @@ func (g *getIter) Last() (*InternalKey, base.LazyValue) {
 func (g *getIter) Next() (*InternalKey, base.LazyValue) {
 	if g.iter != nil {
 		g.iterKey, g.iterValue = g.iter.Next()
+		if err := g.iter.Error(); err != nil {
+			g.err = err
+			return nil, base.LazyValue{}
+		}
 	}
 
 	for {
@@ -134,6 +138,10 @@ func (g *getIter) Next() (*InternalKey, base.LazyValue) {
 				base.InternalKeySeqNumMax,
 			)
 			g.iterKey, g.iterValue = g.iter.SeekGE(g.key, base.SeekGEFlagsNone)
+			if err := g.iter.Error(); err != nil {
+				g.err = err
+				return nil, base.LazyValue{}
+			}
 			g.batch = nil
 			continue
 		}
@@ -151,6 +159,10 @@ func (g *getIter) Next() (*InternalKey, base.LazyValue) {
 			g.rangeDelIter = m.newRangeDelIter(nil)
 			g.mem = g.mem[:n-1]
 			g.iterKey, g.iterValue = g.iter.SeekGE(g.key, base.SeekGEFlagsNone)
+			if err := g.iter.Error(); err != nil {
+				g.err = err
+				return nil, base.LazyValue{}
+			}
 			continue
 		}
 
@@ -181,6 +193,11 @@ func (g *getIter) Next() (*InternalKey, base.LazyValue) {
 					prefix = g.key[:g.comparer.Split(g.key)]
 				}
 				g.iterKey, g.iterValue = g.iter.SeekPrefixGE(prefix, g.key, base.SeekGEFlagsNone)
+				if err := g.iter.Error(); err != nil {
+					g.err = err
+					return nil, base.LazyValue{}
+				}
+
 				if bc.isSyntheticIterBoundsKey || bc.isIgnorableBoundaryKey {
 					g.iterKey = nil
 					g.iterValue = base.LazyValue{}
@@ -219,6 +236,10 @@ func (g *getIter) Next() (*InternalKey, base.LazyValue) {
 			prefix = g.key[:g.comparer.Split(g.key)]
 		}
 		g.iterKey, g.iterValue = g.iter.SeekPrefixGE(prefix, g.key, base.SeekGEFlagsNone)
+		if err := g.iter.Error(); err != nil {
+			g.err = err
+			return nil, base.LazyValue{}
+		}
 		if bc.isSyntheticIterBoundsKey || bc.isIgnorableBoundaryKey {
 			g.iterKey = nil
 			g.iterValue = base.LazyValue{}

--- a/internal/keyspan/interleaving_iter.go
+++ b/internal/keyspan/interleaving_iter.go
@@ -207,11 +207,14 @@ func (i *InterleavingIter) Init(
 	keyspanIter FragmentIterator,
 	opts InterleavingIterOpts,
 ) {
+	keyspanIter = MaybeAssert(keyspanIter, comparer.Compare)
+	// To debug:
+	// keyspanIter = InjectLogging(keyspanIter, base.DefaultLogger)
 	*i = InterleavingIter{
 		cmp:         comparer.Compare,
 		comparer:    comparer,
 		pointIter:   pointIter,
-		keyspanIter: MaybeAssert(keyspanIter, comparer.Compare),
+		keyspanIter: keyspanIter,
 		mask:        opts.Mask,
 		lower:       opts.LowerBound,
 		upper:       opts.UpperBound,

--- a/internal/keyspan/logging_iter.go
+++ b/internal/keyspan/logging_iter.go
@@ -37,7 +37,7 @@ func newLoggingIter(state *loggingState, iter FragmentIterator) FragmentIterator
 	return &loggingIter{
 		iter:    iter,
 		state:   state,
-		context: fmt.Sprintf("%T:", iter),
+		context: fmt.Sprintf("%T(%p):", iter, iter),
 	}
 }
 

--- a/internal/keyspan/logging_iter_test.go
+++ b/internal/keyspan/logging_iter_test.go
@@ -6,6 +6,7 @@ package keyspan
 
 import (
 	"fmt"
+	"regexp"
 	"strings"
 	"testing"
 
@@ -34,7 +35,10 @@ func TestLoggingIter(t *testing.T) {
 			iter = InjectLogging(iter, l)
 			runFragmentIteratorCmd(iter, d.Input, nil)
 			require.NoError(t, iter.Close())
-			return l.String()
+			out := l.String()
+			// Hide pointer values.
+			r := regexp.MustCompile(`\(0x[0-9a-f]+\)`)
+			return r.ReplaceAllString(out, "")
 
 		default:
 			return fmt.Sprintf("unknown command: %s", d.Cmd)

--- a/internal/keyspan/merging_iter.go
+++ b/internal/keyspan/merging_iter.go
@@ -184,7 +184,6 @@ type MergingIter struct {
 	// destination for transforms. Every tranformed span overwrites the
 	// previous.
 	span Span
-	err  error
 	dir  int8
 
 	// alloc preallocates mergingIterLevel and mergingIterItems for use by the
@@ -675,7 +674,7 @@ func (m *MergingIter) Prev() (*Span, error) {
 
 // Close closes the iterator, releasing all acquired resources.
 func (m *MergingIter) Close() error {
-	err := m.err
+	var err error
 	for i := range m.levels {
 		err = firstError(err, m.levels[i].iter.Close())
 	}
@@ -834,7 +833,7 @@ func (m *MergingIter) findNextFragmentSet() (*Span, error) {
 	// below loop will still consider [b,d) before continuing to [d, e)). It
 	// returns when it finds a span that is covered by at least one key.
 
-	for m.heap.len() > 0 && m.err == nil {
+	for m.heap.len() > 0 {
 		// Initialize the next span's start bound. SeekGE and First prepare the
 		// heap without advancing. Next leaves the heap in a state such that the
 		// root is the smallest bound key equal to the returned span's end key,
@@ -876,12 +875,12 @@ func (m *MergingIter) findNextFragmentSet() (*Span, error) {
 		if err := m.nextEntry(); err != nil {
 			return nil, err
 		}
-		for len(m.heap.items) > 0 && m.err == nil && m.cmp(m.heapRoot(), m.start) == 0 {
+		for len(m.heap.items) > 0 && m.cmp(m.heapRoot(), m.start) == 0 {
 			if err := m.nextEntry(); err != nil {
 				return nil, err
 			}
 		}
-		if len(m.heap.items) == 0 || m.err != nil {
+		if len(m.heap.items) == 0 {
 			break
 		}
 
@@ -917,7 +916,7 @@ func (m *MergingIter) findPrevFragmentSet() (*Span, error) {
 	// below loop will still consider [b,d) before continuing to [a, b)). It
 	// returns when it finds a span that is covered by at least one key.
 
-	for m.heap.len() > 0 && m.err == nil {
+	for m.heap.len() > 0 {
 		// Initialize the next span's end bound. SeekLT and Last prepare the
 		// heap without advancing. Prev leaves the heap in a state such that the
 		// root is the largest bound key equal to the returned span's start key,
@@ -955,13 +954,15 @@ func (m *MergingIter) findPrevFragmentSet() (*Span, error) {
 		// L2:          [c, e)
 		// If we're positioned at L1's start(c) start boundary, we want to prev
 		// to move to the first bound < c.
-		m.err = m.prevEntry()
-		for len(m.heap.items) > 0 && m.err == nil && m.cmp(m.heapRoot(), m.end) == 0 {
-			m.err = m.prevEntry()
+		if err := m.prevEntry(); err != nil {
+			return nil, err
 		}
-		if m.err != nil {
-			return nil, m.err
-		} else if len(m.heap.items) == 0 {
+		for len(m.heap.items) > 0 && m.cmp(m.heapRoot(), m.end) == 0 {
+			if err := m.prevEntry(); err != nil {
+				return nil, err
+			}
+		}
+		if len(m.heap.items) == 0 {
 			break
 		}
 

--- a/iterator.go
+++ b/iterator.go
@@ -638,6 +638,12 @@ func (i *Iterator) findNextEntry(limit []byte) {
 			return
 		}
 	}
+
+	// Is iterKey nil due to an error?
+	if err := i.iter.Error(); err != nil {
+		i.err = err
+		i.iterValidityState = IterExhausted
+	}
 }
 
 func (i *Iterator) nextPointCurrentUserKey() bool {
@@ -652,7 +658,15 @@ func (i *Iterator) nextPointCurrentUserKey() bool {
 
 	i.iterKey, i.iterValue = i.iter.Next()
 	i.stats.ForwardStepCount[InternalIterCall]++
-	if i.iterKey == nil || !i.equal(i.key, i.iterKey.UserKey) {
+	if i.iterKey == nil {
+		if err := i.iter.Error(); err != nil {
+			i.err = err
+		} else {
+			i.pos = iterPosNext
+		}
+		return false
+	}
+	if !i.equal(i.key, i.iterKey.UserKey) {
 		i.pos = iterPosNext
 		return false
 	}
@@ -742,8 +756,14 @@ func (i *Iterator) nextUserKey() {
 		i.key = i.keyBuf
 	}
 	for {
-		i.iterKey, i.iterValue = i.iter.Next()
 		i.stats.ForwardStepCount[InternalIterCall]++
+		i.iterKey, i.iterValue = i.iter.Next()
+		if i.iterKey == nil {
+			if err := i.iter.Error(); err != nil {
+				i.err = err
+				return
+			}
+		}
 		// NB: We're guaranteed to be on the next user key if the previous key
 		// had a zero sequence number (`done`), or the new key has a trailer
 		// greater or equal to the previous key's trailer. This is true because
@@ -952,6 +972,13 @@ func (i *Iterator) findPrevEntry(limit []byte) {
 			// the key prefix.
 			i.stats.ReverseStepCount[InternalIterCall]++
 			i.iterKey, i.iterValue = i.iter.Prev()
+			if i.iterKey == nil {
+				if err := i.iter.Error(); err != nil {
+					i.err = err
+					i.iterValidityState = IterExhausted
+					return
+				}
+			}
 			if limit != nil && i.iterKey != nil && i.cmp(limit, i.iterKey.UserKey) > 0 && !i.rangeKeyWithinLimit(limit) {
 				i.iterValidityState = IterAtLimit
 				i.pos = iterPosCurReversePaused
@@ -992,8 +1019,8 @@ func (i *Iterator) findPrevEntry(limit []byte) {
 			i.value = LazyValue{}
 			i.iterValidityState = IterExhausted
 			valueMerger = nil
-			i.iterKey, i.iterValue = i.iter.Prev()
 			i.stats.ReverseStepCount[InternalIterCall]++
+			i.iterKey, i.iterValue = i.iter.Prev()
 			// Compare with the limit. We could optimize by only checking when
 			// we step to the previous user key, but detecting that requires a
 			// comparison too. Note that this position may already passed a
@@ -1051,12 +1078,14 @@ func (i *Iterator) findPrevEntry(limit []byte) {
 					i.lazyValueBuf = value[:0]
 				}
 				if i.err != nil {
+					i.iterValidityState = IterExhausted
 					return
 				}
 				valueMerger, i.err = i.merge(i.key, value)
 				var iterValue []byte
 				iterValue, _, i.err = i.iterValue.Value(nil)
 				if i.err != nil {
+					i.iterValidityState = IterExhausted
 					return
 				}
 				if i.err == nil {
@@ -1070,6 +1099,7 @@ func (i *Iterator) findPrevEntry(limit []byte) {
 				var iterValue []byte
 				iterValue, _, i.err = i.iterValue.Value(nil)
 				if i.err != nil {
+					i.iterValidityState = IterExhausted
 					return
 				}
 				i.err = valueMerger.MergeNewer(iterValue)
@@ -1129,6 +1159,10 @@ func (i *Iterator) prevUserKey() {
 		i.iterKey, i.iterValue = i.iter.Prev()
 		i.stats.ReverseStepCount[InternalIterCall]++
 		if i.iterKey == nil {
+			if err := i.iter.Error(); err != nil {
+				i.err = err
+				i.iterValidityState = IterExhausted
+			}
 			break
 		}
 		if !i.equal(i.key, i.iterKey.UserKey) {
@@ -1327,6 +1361,11 @@ func (i *Iterator) SeekGEWithLimit(key []byte, limit []byte) IterValidityState {
 	if seekInternalIter {
 		i.iterKey, i.iterValue = i.iter.SeekGE(key, flags)
 		i.stats.ForwardSeekCount[InternalIterCall]++
+		if err := i.iter.Error(); err != nil {
+			i.err = err
+			i.iterValidityState = IterExhausted
+			return i.iterValidityState
+		}
 	}
 	i.findNextEntry(limit)
 	i.maybeSampleRead()
@@ -1581,6 +1620,11 @@ func (i *Iterator) SeekLTWithLimit(key []byte, limit []byte) IterValidityState {
 	if seekInternalIter {
 		i.iterKey, i.iterValue = i.iter.SeekLT(key, base.SeekLTFlagsNone)
 		i.stats.ReverseSeekCount[InternalIterCall]++
+		if err := i.iter.Error(); err != nil {
+			i.err = err
+			i.iterValidityState = IterExhausted
+			return i.iterValidityState
+		}
 	}
 	i.findPrevEntry(limit)
 	i.maybeSampleRead()
@@ -1619,6 +1663,10 @@ func (i *Iterator) First() bool {
 	i.stats.ForwardSeekCount[InterfaceCall]++
 
 	i.iterFirstWithinBounds()
+	if i.err != nil {
+		i.iterValidityState = IterExhausted
+		return false
+	}
 	i.findNextEntry(nil)
 	i.maybeSampleRead()
 	return i.iterValidityState == IterValid
@@ -1651,6 +1699,10 @@ func (i *Iterator) Last() bool {
 	i.stats.ReverseSeekCount[InterfaceCall]++
 
 	i.iterLastWithinBounds()
+	if i.err != nil {
+		i.iterValidityState = IterExhausted
+		return false
+	}
 	i.findPrevEntry(nil)
 	i.maybeSampleRead()
 	return i.iterValidityState == IterValid
@@ -2013,9 +2065,15 @@ func (i *Iterator) PrevWithLimit(limit []byte) IterValidityState {
 			i.iterLastWithinBounds()
 		} else {
 			i.prevUserKey()
+			if i.err != nil {
+				return i.iterValidityState
+			}
 		}
 		if stepAgain {
 			i.prevUserKey()
+			if i.err != nil {
+				return i.iterValidityState
+			}
 		}
 	}
 	i.findPrevEntry(limit)
@@ -2209,6 +2267,7 @@ func (i *Iterator) ValueAndErr() ([]byte, error) {
 	val, callerOwned, err := i.value.Value(i.lazyValueBuf)
 	if err != nil {
 		i.err = err
+		i.iterValidityState = IterExhausted
 	}
 	if callerOwned {
 		i.lazyValueBuf = val[:0]
@@ -2238,7 +2297,7 @@ func (i *Iterator) Valid() bool {
 	valid := i.iterValidityState == IterValid && !i.requiresReposition
 	if invariants.Enabled {
 		if err := i.Error(); valid && err != nil {
-			panic(errors.WithSecondaryError(errors.AssertionFailedf("pebble: iterator is valid with non-nil Error"), err))
+			panic(errors.AssertionFailedf("pebble: iterator is valid with non-nil Error: %+v", err))
 		}
 	}
 	return valid
@@ -2246,10 +2305,13 @@ func (i *Iterator) Valid() bool {
 
 // Error returns any accumulated error.
 func (i *Iterator) Error() error {
-	if i.iter != nil {
-		return firstError(i.err, i.iter.Error())
+	if i.err != nil {
+		return i.err
 	}
-	return i.err
+	if i.iter != nil {
+		return i.iter.Error()
+	}
+	return nil
 }
 
 const maxKeyBufCacheSize = 4 << 10 // 4 KB

--- a/metamorphic/test.go
+++ b/metamorphic/test.go
@@ -331,11 +331,11 @@ func (t *Test) maybeSaveData() {
 	}
 }
 
-// Step runs one single operation, returning whether or not there are additional
-// operations, the operation's output and an error if any occurred while running
-// the operation.
+// Step runs one single operation, returning: whether there are additional
+// operations remaining; the operation's output; and an error if any occurred
+// while running the operation.
 //
-// Step may be used in stead of Execute to advance a test one operation at a
+// Step may be used instead of Execute to advance a test one operation at a
 // time.
 func (t *Test) Step() (more bool, operationOutput string, err error) {
 	more = t.step(t.h, func(format string, args ...interface{}) {

--- a/vfs/errorfs/errorfs.go
+++ b/vfs/errorfs/errorfs.go
@@ -223,17 +223,15 @@ func (c *Counter) MaybeError(op Op) error {
 // Load returns the number of errors injected.
 func (c *Counter) Load() uint64 {
 	c.mu.Lock()
-	v := c.mu.v
-	c.mu.Unlock()
-	return v
+	defer c.mu.Unlock()
+	return c.mu.v
 }
 
 // LastError returns the last non-nil error injected.
 func (c *Counter) LastError() error {
 	c.mu.Lock()
-	err := c.mu.lastErr
-	c.mu.Unlock()
-	return err
+	defer c.mu.Unlock()
+	return c.mu.lastErr
 }
 
 // Toggle wraps an Injector. By default, Toggle injects nothing. When toggled on

